src/rpc/server.zig

```diff
@@ -0,0 +1,547 @@
+                const full_archive_name = full_archive_name_bounded.constSlice();
+                if (std.mem.eql(u8, path, full_archive_name)) {
+                    const archive_file = try snapshot_dir.openFile(full_archive_name, .{});
+                    defer archive_file.close();
+                    var send_buffer: [4096]u8 = undefined;
+                    try httpResponseSendFile(request, archive_file, &send_buffer);
+                    return;
+                }
+            }
+
+            if (inc_info) |inc| {
+                const inc_archive_name_bounded = inc.snapshotNameStr();
+                const inc_archive_name = inc_archive_name_bounded.constSlice();
+                if (std.mem.eql(u8, path, inc_archive_name)) {
+                    const archive_file = try snapshot_dir.openFile(inc_archive_name, .{});
+                    defer archive_file.close();
+                    var send_buffer: [4096]u8 = undefined;
+                    try httpResponseSendFile(request, archive_file, &send_buffer);
+                    return;
+                }
+            }
+        },
+        else => {},
+    }
+
+    logger.err().logf(
+        "{} made an unrecognized request '{} {s}'",
+        .{ conn_address, methodFmt(request.head.method), request.head.target },
+    );
+    try request.respond("", .{
+        .status = .not_found,
+        .keep_alive = false,
+    });
+}
+
+fn httpResponseSendFile(
+    request: *std.http.Server.Request,
+    archive_file: std.fs.File,
+    send_buffer: []u8,
+) !void {
+    const archive_len = try archive_file.getEndPos();
+
+    var response = request.respondStreaming(.{
+        .send_buffer = send_buffer,
+        .content_length = archive_len,
+    });
+    const writer = sig.utils.io.narrowAnyWriter(
+        response.writer(),
+        std.http.Server.Response.WriteError,
+    );
+
+    const Fifo = std.fifo.LinearFifo(u8, .{ .Static = 1 });
+    var fifo: Fifo = Fifo.init();
+    try archive_file.seekTo(0);
+    try fifo.pump(archive_file.reader(), writer);
+
+    try response.end();
+}
+
+const AcceptConnectionError = error{
+    ProcessFdQuotaExceeded,
+    SystemFdQuotaExceeded,
+    SystemResources,
+    ProtocolFailure,
+    BlockedByFirewall,
+    NetworkSubsystemFailed,
+} || std.posix.UnexpectedError;
+fn acceptConnection(
+    tcp_server: *std.net.Server,
+    logger: ScopedLogger,
+) AcceptConnectionError!?std.net.Server.Connection {
+    const conn = tcp_server.accept() catch |err| switch (err) {
+        error.Unexpected,
+        => |e| return e,
+
+        error.ProcessFdQuotaExceeded,
+        error.SystemFdQuotaExceeded,
+        error.SystemResources,
+        error.ProtocolFailure,
+        error.BlockedByFirewall,
+        error.NetworkSubsystemFailed,
+        => |e| return e,
+
+        error.FileDescriptorNotASocket,
+        error.SocketNotListening,
+        error.OperationNotSupported,
+        => unreachable, // Improperly initialized server.
```

Comment:
The only circumstance under which these errors ever occur is if you initialize a `Server` directly with bogus values, ie, intentionally misusing the API, which is something that would be pretty easily flagged in review; I don't really think it's worth the cost, but I'll change this if you're set in that position.

Reply:
Whats the cost? I think we should we should stick to the approach of only using unreachable when a given path is **provably** unreachable

Reply:
@panic("Improperly initialised server") is also self commenting code

Reply:
The cost is the branch; an unreachable will be compiled out, a panic forces another code path, even if that branch never actually occurs.
"Provably unreachable" is a bit nebulous, since the most literal interpretation just means it can be encoded more simply directly in the type system, eschewing the need for `unreachable`; for all sane intents and purposes, this code is provably unreachable, because it will always be a socket, it will always be listening, and the operation will be supported.

Reply:
"Provably unreachable" came directly from our chat recently in sig channel, I get that it compiles out the branch but is this actually a considerable cost in this instance? Genuinely curious here

Reply:
It just seems like a pointless cost to pay at all, it's comparable to writing `ptr orelse @panic("null pointer")` instead of `ptr.?` after branching on `ptr != null`; if the handle came from some generic source that we ought to guard against carefully, I'd agree with your assessment, but in this case the code is in full control of the handle.

Reply:
agreed with harold in this case - if its improperly initialized then we should panic (as the 'unreachable' is actually reachable) given the server is a parameter - this isnt performance sensitive code, one branch is okay imo

Reply:
3761d42d41ee276fdc40dfb74e0499d34ac74249